core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,751 @@
+					uint32_t cmd,
+					struct tee_ta_param *param,
+					TEE_ErrorOrigin *eo __unused)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	TEE_Result tmp_res = TEE_SUCCESS;
+	unsigned int ns_buf_size = 0;
+	struct param_mem *mem = NULL;
+	void *va = NULL;
+
+	if (cmd != PTA_STMM_COMMUNICATE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mem = &param->u[0].mem;
+	ns_buf_size = mem->size;
+	if (ns_buf_size > spc->ns_comm_buf_size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	res = mobj_inc_map(mem->mobj);
+	if (res)
+		return res;
+
+	va = mobj_get_va(mem->mobj, mem->offs);
+	if (!va) {
+		EMSG("Can't get a valid VA for NS buffer\n");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	spc->regs.x[0] = FFA_MSG_SEND_DIRECT_REQ_64;
+	spc->regs.x[1] = (stmm_pta_id << 16) | stmm_id;
+	spc->regs.x[2] = FFA_PARAM_MBZ;
+	spc->regs.x[3] = spc->ns_comm_buf_addr;
+	spc->regs.x[4] = ns_buf_size;
+	spc->regs.x[5] = 0;
+	spc->regs.x[6] = 0;
+	spc->regs.x[7] = 0;
+
+	tee_ta_push_current_session(s);
+
+	memcpy((void *)spc->ns_comm_buf_addr, va, ns_buf_size);
+
+	res = sec_part_enter_user_mode(spc);
+	/*
+	 * Copy the SPM response from secure partiton back to the non secure
+	 * the client that called us
+	 */
+	param->u[1].val.a = spc->regs.x[4];
+	if (res)
+		goto out;
+
+	memcpy(va, (void *)spc->ns_comm_buf_addr, ns_buf_size);
+
+out:
+	tmp_res = mobj_dec_map(mem->mobj);
+	assert(!tmp_res);
+	tee_ta_pop_current_session();
+
+	return res;
+}
+
+static void stmm_enter_close_session(struct tee_ta_session *s __unused)
+{
+}
+
+static void sec_part_dump_state(struct tee_ta_ctx *ctx)
+{
+	user_mode_ctx_print_mappings(to_user_mode_ctx(ctx));
+}
+
+static uint32_t sec_part_get_instance_id(struct tee_ta_ctx *ctx)
+{
+	return to_sec_part_ctx(ctx)->uctx.vm_info.asid;
+}
+
+static void sec_part_ctx_destroy(struct tee_ta_ctx *ctx)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(ctx);
+
+	tee_pager_rem_um_areas(&spc->uctx);
+	vm_info_final(&spc->uctx);
+	free(spc);
+}
+
+static uint32_t sp_svc_get_mem_attr(vaddr_t va)
+{
+	TEE_Result res = TEE_ERROR_BAD_PARAMETERS;
+	struct tee_ta_session *sess = NULL;
+	struct sec_part_ctx *spc = NULL;
+	uint16_t attrs = 0;
+	uint16_t perm = 0;
+
+	if (!va)
+		goto out;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	spc = to_sec_part_ctx(sess->ctx);
+
+	res = vm_get_prot(&spc->uctx, va, SMALL_PAGE_SIZE, &attrs);
+	if (res)
+		goto out;
+
+	if (attrs & TEE_MATTR_UR)
+		perm |= SP_MEM_ATTR_ACCESS_RO;
+	else if (attrs & TEE_MATTR_UW)
+		perm |= SP_MEM_ATTR_ACCESS_RW;
+
+	if (attrs & TEE_MATTR_UX)
+		perm |= SP_MEM_ATTR_EXEC;
+
+	return perm;
+out:
+	return SP_RET_DENIED;
+}
+
+static int sp_svc_set_mem_attr(vaddr_t va, unsigned int nr_pages, uint32_t perm)
+{
+	TEE_Result res = TEE_ERROR_BAD_PARAMETERS;
+	struct tee_ta_session *sess = NULL;
+	struct sec_part_ctx *spc = NULL;
+	uint32_t prot = 0;
+
+	if (!va || !nr_pages)
```

Comment:
Should we check that unused bits in `perm` are `0`?

Reply:
As far I've seen the StMM is calling ArmSetMemoryRegionNoExec, ArmClearMemoryRegionNoExec, ArmSetMemoryRegionReadOnly, ArmClearMemoryRegionReadOnly (which actually sets it as RW).
I did print the bits before since it's ORing the new requested attributes with the existing ones and the unused bits were always 0. 
I guess I can add it, since it won't break anything and since we depend on running an application we don't have control over, being explicit about the bits might save us from future bugs